File: app/src/main/java/dev/ragnarok/fenrir/fragment/fave/FaveTabsFragment.java
package dev.ragnarok.fenrir.fragment.fave;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.tabs.TabLayoutMediator;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.fenrir.Extra;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.activity.ActivityFeatures;
import dev.ragnarok.fenrir.activity.ActivityUtils;
import dev.ragnarok.fenrir.fragment.AbsNavigationFragment;
import dev.ragnarok.fenrir.fragment.base.BaseFragment;
import dev.ragnarok.fenrir.link.types.FaveLink;
import dev.ragnarok.fenrir.listener.OnSectionResumeCallback;
import dev.ragnarok.fenrir.place.Place;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.Utils;

public class FaveTabsFragment extends BaseFragment {

    public static final int TAB_UNKNOWN = -1;
    public static final int TAB_PAGES = 0;
    public static final int TAB_GROUPS = 1;
    public static final int TAB_POSTS = 2;
    public static final int TAB_LINKS = 3;
    public static final int TAB_ARTICLES = 4;
    public static final int TAB_PRODUCTS = 5;
    public static final int TAB_PHOTOS = 6;
    public static final int TAB_VIDEOS = 7;
    private int mAccountId;

    public static Bundle buildArgs(int accountId, int tab) {
        Bundle args = new Bundle();
        args.putInt(Extra.ACCOUNT_ID, accountId);
        args.putInt(Extra.TAB, tab);
        return args;
    }

    public static FaveTabsFragment newInstance(int accountId, int tab) {
        return newInstance(buildArgs(accountId, tab));
    }

    public static FaveTabsFragment newInstance(Bundle args) {
        FaveTabsFragment faveTabsFragment = new FaveTabsFragment();
        faveTabsFragment.setArguments(args);
        return faveTabsFragment;
    }

    public static int getTabByLinkSection(String linkSection) {
        if (TextUtils.isEmpty(linkSection)) {
            return TAB_PHOTOS;
        }

        switch (linkSection) {
            case FaveLink.SECTION_PHOTOS:
                return TAB_PHOTOS;
            case FaveLink.SECTION_VIDEOS:
                return TAB_VIDEOS;
            case FaveLink.SECTION_POSTS:
                return TAB_POSTS;
            case FaveLink.SECTION_PAGES:
                return TAB_PAGES;
            case FaveLink.SECTION_LINKS:
                return TAB_LINKS;
            case FaveLink.SECTION_ARTICLES:
                return TAB_ARTICLES;
            case FaveLink.SECTION_PRODUCTS:
                return TAB_PRODUCTS;
            default:
                return TAB_UNKNOWN;
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mAccountId = requireArguments().getInt(Extra.ACCOUNT_ID);
    }

    public int getAccountId() {
        return mAccountId;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        ViewGroup root = (ViewGroup) inflater.inflate(R.layout.fragment_fave_tabs, container, false);
        ((AppCompatActivity) requireActivity()).setSupportActionBar(root.findViewById(R.id.toolbar));
        return root;
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        ViewPager2 viewPager = view.findViewById(R.id.viewpager);
        viewPager.setOffscreenPageLimit(1);
        viewPager.setPageTransformer(Utils.createPageTransform(Settings.get().main().getViewpager_page_transform()));
        setupViewPager(viewPager, view);

        if (requireArguments().containsKey(Extra.TAB)) {
            int tab = requireArguments().getInt(Extra.TAB);
            requireArguments().remove(Extra.TAB);
            viewPager.setCurrentItem(tab);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        Settings.get().ui().notifyPlaceResumed(Place.BOOKMARKS);

        ActionBar actionBar = ActivityUtils.supportToolbarFor(this);

        if (actionBar != null) {
            actionBar.setTitle(R.string.bookmarks);
            actionBar.setSubtitle(null);
        }

        if (requireActivity() instanceof OnSectionResumeCallback) {
            ((OnSectionResumeCallback) requireActivity()).onSectionResume(AbsNavigationFragment.SECTION_ITEM_BOOKMARKS);
        }

        new ActivityFeatures.Builder()
                .begin()
                .setHideNavigationMenu(false)
                .setBarsColored(requireActivity(), true)
                .build()
                .apply(requireActivity());
    }

    private void setupViewPager(ViewPager2 viewPager, @NonNull View view) {
        List<ITab> tabs = new ArrayList<>();
        tabs.add(new Tab(() -> FavePagesFragment.newInstance(getAccountId(), true), getString(R.string.pages)));
        tabs.add(new Tab(() -> FavePagesFragment.newInstance(getAccountId(), false), getString(R.string.groups)));
        tabs.add(new Tab(() -> FavePostsFragment.newInstance(getAccountId()), getString(R.string.posts)));
        tabs.add(new Tab(() -> FaveLinksFragment.newInstance(getAccountId()), getString(R.string.links)));
        tabs.add(new Tab(() -> FaveArticlesFragment.newInstance(getAccountId()), getString(R.string.articles)));
        tabs.add(new Tab(() -> FaveProductsFragment.newInstance(getAccountId()), getString(R.string.products)));
        tabs.add(new Tab(() -> FavePhotosFragment.newInstance(getAccountId()), getString(R.string.photos)));
        tabs.add(new Tab(() -> FaveVideosFragment.newInstance(getAccountId()), getString(R.string.videos)));
        Adapter adapter = new Adapter(tabs, this);
        viewPager.setAdapter(adapter);

        new TabLayoutMediator(view.findViewById(R.id.tablayout), viewPager, (tab, position) ->
                tab.setText(tabs.get(position).getTabTitle())).attach();
    }

    private interface ITab {
        String getTabTitle();

        IFragmentCreator getFragmentCreator();
    }

    private interface IFragmentCreator {
        Fragment create();
    }

    private static class Tab implements ITab {

        final String title;
        final IFragmentCreator creator;

        private Tab(IFragmentCreator creator, String title) {
            this.title = title;
            this.creator = creator;
        }

        @Override
        public String getTabTitle() {
            return title;
        }

        @Override
        public IFragmentCreator getFragmentCreator() {
            return creator;
        }
    }

    static class Adapter extends FragmentStateAdapter {

        private final List<ITab> tabs;

        public Adapter(List<ITab> tabs, @NonNull Fragment fragmentActivity) {
            super(fragmentActivity);
            this.tabs = tabs;
        }

        @NonNull
        @Override
        public Fragment createFragment(int position) {
            return tabs.get(position).getFragmentCreator().create();
        }

        @Override
        public int getItemCount() {
            return tabs.size();
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/fave/FaveTabsFragment.kt
package dev.ragnarok.fenrir.fragment.fave

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import dev.ragnarok.fenrir.Extra
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.activity.ActivityFeatures
import dev.ragnarok.fenrir.activity.ActivityUtils.supportToolbarFor
import dev.ragnarok.fenrir.fragment.base.BaseFragment
import dev.ragnarok.fenrir.fragment.fave.favearticles.FaveArticlesFragment
import dev.ragnarok.fenrir.fragment.fave.favelinks.FaveLinksFragment
import dev.ragnarok.fenrir.fragment.fave.favepages.FavePagesFragment.Companion.newInstance
import dev.ragnarok.fenrir.fragment.fave.favephotos.FavePhotosFragment
import dev.ragnarok.fenrir.fragment.fave.faveposts.FavePostsFragment
import dev.ragnarok.fenrir.fragment.fave.faveproducts.FaveProductsFragment
import dev.ragnarok.fenrir.fragment.fave.favevideos.FaveVideosFragment
import dev.ragnarok.fenrir.link.types.FaveLink
import dev.ragnarok.fenrir.listener.OnSectionResumeCallback
import dev.ragnarok.fenrir.place.Place
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.Utils.createPageTransform
import dev.ragnarok.fenrir.view.navigation.AbsNavigationView

class FaveTabsFragment : BaseFragment() {
    var accountId = 0L
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        accountId = requireArguments().getLong(Extra.ACCOUNT_ID)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val root = inflater.inflate(R.layout.fragment_fave_tabs, container, false) as ViewGroup
        (requireActivity() as AppCompatActivity).setSupportActionBar(root.findViewById(R.id.toolbar))
        return root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val viewPager: ViewPager2 = view.findViewById(R.id.viewpager)
        viewPager.offscreenPageLimit = 1
        viewPager.setPageTransformer(
            createPageTransform(
                Settings.get().main().viewpager_page_transform
            )
        )
        setupViewPager(viewPager, view)
        if (requireArguments().containsKey(Extra.TAB)) {
            val tab = requireArguments().getInt(Extra.TAB)
            requireArguments().remove(Extra.TAB)
            viewPager.currentItem = tab
        }
    }

    override fun onResume() {
        super.onResume()
        Settings.get().ui().notifyPlaceResumed(Place.BOOKMARKS)
        val actionBar = supportToolbarFor(this)
        if (actionBar != null) {
            actionBar.setTitle(R.string.bookmarks)
            actionBar.subtitle = null
        }
        if (requireActivity() is OnSectionResumeCallback) {
            (requireActivity() as OnSectionResumeCallback).onSectionResume(AbsNavigationView.SECTION_ITEM_BOOKMARKS)
        }
        ActivityFeatures.Builder()
            .begin()
            .setHideNavigationMenu(false)
            .setBarsColored(requireActivity(), true)
            .build()
            .apply(requireActivity())
    }

    private fun setupViewPager(viewPager: ViewPager2, view: View) {
        val tabs: MutableList<ITab> = ArrayList()
        tabs.add(Tab(object : IFragmentCreator {
            override fun create(): Fragment {
                return newInstance(
                    accountId, true
                )
            }
        }, getString(R.string.pages)))
        tabs.add(Tab(object : IFragmentCreator {
            override fun create(): Fragment {
                return newInstance(
                    accountId, false
                )
            }
        }, getString(R.string.groups)))
        tabs.add(Tab(object : IFragmentCreator {
            override fun create(): Fragment {
                return FavePostsFragment.newInstance(
                    accountId
                )
            }
        }, getString(R.string.posts)))
        tabs.add(Tab(object : IFragmentCreator {
            override fun create(): Fragment {
                return FaveLinksFragment.newInstance(
                    accountId
                )
            }
        }, getString(R.string.links)))
        tabs.add(Tab(object : IFragmentCreator {
            override fun create(): Fragment {
                return FaveArticlesFragment.newInstance(
                    accountId
                )
            }
        }, getString(R.string.articles)))
        tabs.add(Tab(object : IFragmentCreator {
            override fun create(): Fragment {
                return FaveProductsFragment.newInstance(
                    accountId
                )
            }
        }, getString(R.string.products)))
        tabs.add(Tab(object : IFragmentCreator {
            override fun create(): Fragment {
                return FavePhotosFragment.newInstance(
                    accountId
                )
            }
        }, getString(R.string.photos)))
        tabs.add(Tab(object : IFragmentCreator {
            override fun create(): Fragment {
                return FaveVideosFragment.newInstance(
                    accountId
                )
            }
        }, getString(R.string.videos)))
        val adapter = Adapter(tabs, this)
        viewPager.adapter = adapter
        TabLayoutMediator(
            view.findViewById(R.id.tablayout),
            viewPager
        ) { tab: TabLayout.Tab, position: Int -> tab.text = tabs[position].tabTitle }.attach()
    }

    interface ITab {
        val tabTitle: String?
        val fragmentCreator: IFragmentCreator
    }

    interface IFragmentCreator {
        fun create(): Fragment
    }

    private class Tab(override val fragmentCreator: IFragmentCreator, val title: String) : ITab {
        override val tabTitle: String
            get() = title
    }

    internal class Adapter(private val tabs: List<ITab>, fragmentActivity: Fragment) :
        FragmentStateAdapter(fragmentActivity) {
        override fun createFragment(position: Int): Fragment {
            return tabs[position].fragmentCreator.create()
        }

        override fun getItemCount(): Int {
            return tabs.size
        }
    }

    companion object {
        const val TAB_UNKNOWN = -1
        const val TAB_PAGES = 0
        const val TAB_GROUPS = 1
        private const val TAB_POSTS = 2
        private const val TAB_LINKS = 3
        private const val TAB_ARTICLES = 4
        private const val TAB_PRODUCTS = 5
        private const val TAB_PHOTOS = 6
        private const val TAB_VIDEOS = 7
        fun buildArgs(accountId: Long, tab: Int): Bundle {
            val args = Bundle()
            args.putLong(Extra.ACCOUNT_ID, accountId)
            args.putInt(Extra.TAB, tab)
            return args
        }

        fun newInstance(accountId: Long, tab: Int): FaveTabsFragment {
            return newInstance(buildArgs(accountId, tab))
        }

        fun newInstance(args: Bundle?): FaveTabsFragment {
            val faveTabsFragment = FaveTabsFragment()
            faveTabsFragment.arguments = args
            return faveTabsFragment
        }

        fun getTabByLinkSection(linkSection: String?): Int {
            return if (linkSection.isNullOrEmpty()) {
                TAB_PHOTOS
            } else when (linkSection) {
                FaveLink.SECTION_PHOTOS -> TAB_PHOTOS
                FaveLink.SECTION_VIDEOS -> TAB_VIDEOS
                FaveLink.SECTION_POSTS -> TAB_POSTS
                FaveLink.SECTION_PAGES -> TAB_PAGES
                FaveLink.SECTION_LINKS -> TAB_LINKS
                FaveLink.SECTION_ARTICLES -> TAB_ARTICLES
                FaveLink.SECTION_PRODUCTS -> TAB_PRODUCTS
                else -> TAB_UNKNOWN
            }
        }
    }
}